update docs for new campaign flow (#10921)
sourcegraph/sourcegraph: doc/dev/campaigns_design.md

# Campaigns design doc

Why are [campaigns](../user/campaigns/index.md) designed the way they are?

## Principles

> - **Declarative API** (not imperative). You declare your intent, such as "lint files in all repositories with a `package.json` file"<!-- TODO(sqs): thorsten had a suggestion to make this quote use non-imperative language but I can't find the comment -->. The campaign figures out how to achieve your desired state. The external state (of repositories, changesets, code hosts, access tokens, etc.) can change at any time, and temporary errors frequently occur when reading and writing to code hosts. These factors would make an imperative API very cumbersome because each API client would need to handle the complexity of the distributed system.

Reviewer: ```suggestion
- **Declarative API** (not imperative). You declare your intent, such as "files in all repositories with a `package.json` file should be linted". The campaign figures out how to achieve your desired state. The external state (of repositories, changesets, code hosts, access tokens, etc.) can change at any time, and temporary errors frequently occur when reading and writing to code hosts. These factors would make an imperative API very cumbersome because each API client would need to handle the complexity of the distributed system.
```
